File: Sharebook/Server/src/Services/Profile/Sharebook.Profile.API/DTO/RelationshipStateDTO.cs
namespace Sharebook.Profile.DTO
{
    using System.Collections.Generic;
    using Newtonsoft.Json;

    using Sharebook.Common.Public.Models.Interfaces.DTO;

    public class RelationshipStateDTO : IRelationshipStateDTO
    {
        private List<string> connections;

        [JsonProperty("id")]
        public long TargetUserId { get; set; }

        [JsonProperty("id_str")]
        public string TargetUserIdStr { get; set; }

        [JsonProperty("name")]
        public string TargetUserName { get; set; }

        [JsonProperty("screen_name")]
        public string TargetUserScreenName { get; set; }

        [JsonIgnore]
        public bool Following { get; set; }
        
        [JsonIgnore]
        public bool FollowedBy { get; set; }

        [JsonIgnore]
        public bool FollowingRequested { get; set; }

        [JsonIgnore]
        public bool FollowingRequestReceived { get; set; }

        [JsonProperty("connections")]
        public List<string> Connections
        {
            get => this.connections;
            set
            {
                this.connections = value;
                UpdateConnectionInfos();
            }
        }
        
        private void UpdateConnectionInfos()
        {
            Following = this.connections.Contains("following");
            FollowedBy = this.connections.Contains("followed_by");
            FollowingRequested = this.connections.Contains("following_requested");
            FollowingRequestReceived = this.connections.Contains("following_received");
        }
    }
}

File: Sharebook/Client/src/app/shared/relationships/RelationshipStateDTO.ts
import {IRelationshipStateDTO} from "./IRelationshipStateDTO";

export class RelationshipStateDTO implements IRelationshipStateDTO {
  private _connections: Array<string>;

  // [JsonProperty("id")]
  public targetUserId: number;

  // [JsonProperty("id_str")]
  public targetUserIdStr: string;

  // [JsonProperty("name")]
  public targetUserName: string;

  // [JsonProperty("screen_name")]
  public targetUserScreenName: string;

  // [JsonIgnore]
  public following: boolean;

  // [JsonIgnore]
  public followedBy: boolean;

  // [JsonIgnore]
  public followingRequested: boolean;

  // [JsonIgnore]
  public followingRequestReceived: boolean;

  // [JsonProperty("connections")]
  get connections(): Array<string> {
    return this._connections;
  }

  set connections(value: Array<string>) {
    this._connections = value;
    this.UpdateConnectionInfos();
  }

  private UpdateConnectionInfos(): void {
    this.following = this._connections.includes("following");
    this.followedBy = this._connections.includes("followed_by");
    this.followingRequested = this._connections.includes("following_requested");
    this.followingRequestReceived = this._connections.includes("following_received");
  }
}
